File: ios/liquid_toasts/Sources/liquid_toasts/DeadlineScheduler.swift
import Foundation

/// Owns all auto-dismiss timing for the toast stack: wall-clock deadlines
/// (so backgrounding can't corrupt them), the `Task`s that fire them, and the
/// banked remaining time for toasts paused mid-interaction.
///
/// Deliberately knows nothing about the stack itself — the manager guards
/// existence and reacts to `onExpire`. Also deliberately Flutter-free, so it
/// could be unit-tested from a Flutter-free subtarget later.
@MainActor
final class DeadlineScheduler {
  /// Fired when a deadline expires. `reason` is `"timeout"` (timer fired live)
  /// or `"appBackgrounded"` (found past-due while foregrounding).
  var onExpire: ((_ id: String, _ reason: String) -> Void)?

  /// Wall-clock deadlines. Survive backgrounding; the tasks that watch them
  /// are cancelled in the background and rebuilt on foreground.
  private var deadlines: [String: Date] = [:]
  private var tasks: [String: Task<Void, Never>] = [:]
  /// Banked remaining time for toasts paused mid-interaction (touch-down).
  private var pausedRemaining: [String: TimeInterval] = [:]
  private var backgrounded = false

  /// Arms (or re-arms) the auto-dismiss for [id]. `nil` disarms. A fresh arm
  /// supersedes any banked pause.
  func arm(id: String, duration: TimeInterval?) {
    cancelTask(id)
    pausedRemaining[id] = nil
    guard let duration else {
      deadlines[id] = nil
      return
    }
    let deadline = Date().addingTimeInterval(duration)
    deadlines[id] = deadline
    if !backgrounded { schedule(id: id, fireAt: deadline) }
  }

  /// Cancels everything known about [id] (teardown, async-action spinner).
  func disarm(id: String) {
    cancelTask(id)
    deadlines[id] = nil
    pausedRemaining[id] = nil
  }

  /// Pauses [id]'s auto-dismiss while the user is touching it: banks the
  /// remaining time and clears the wall-clock deadline so neither the timer
  /// nor a background/foreground cycle can fire mid-interaction. No-op when
  /// there is no live deadline (persistent / loading toasts).
  func pause(id: String) {
    guard let deadline = deadlines[id] else { return }
    let remaining = deadline.timeIntervalSinceNow
    guard remaining > 0 else { return }
    pausedRemaining[id] = remaining
    cancelTask(id)
    deadlines[id] = nil
  }

  /// Resumes a paused [id] with its banked remaining time. No-op if it was
  /// never paused.
  func resume(id: String) {
    guard let remaining = pausedRemaining.removeValue(forKey: id) else { return }
    let deadline = Date().addingTimeInterval(remaining)
    deadlines[id] = deadline
    if !backgrounded { schedule(id: id, fireAt: deadline) }
  }

  /// Hot restart / flush: drop all timing state.
  func removeAll() {
    for task in tasks.values { task.cancel() }
    tasks.removeAll()
    deadlines.removeAll()
    pausedRemaining.removeAll()
  }

  // MARK: - App lifecycle

  /// Cancels the watcher tasks but keeps the wall-clock deadlines — they are
  /// re-evaluated against real time on foreground.
  func appDidEnterBackground() {
    backgrounded = true
    for task in tasks.values { task.cancel() }
    tasks.removeAll()
  }

  /// Sweeps the stored deadlines: past-due ones expire immediately (reason
  /// `"appBackgrounded"`), live ones get fresh watcher tasks.
  func appWillEnterForeground() {
    backgrounded = false
    let now = Date()
    for (id, deadline) in deadlines {
      if deadline <= now {
        deadlines[id] = nil
        onExpire?(id, "appBackgrounded")
      } else {
        schedule(id: id, fireAt: deadline)
      }
    }
  }

  // MARK: - Internals

  private func schedule(id: String, fireAt: Date) {
    tasks[id]?.cancel()
    tasks[id] = Task { [weak self] in
      let interval = max(0, fireAt.timeIntervalSinceNow)
      try? await Task.sleep(for: .seconds(interval))
      guard !Task.isCancelled, let self, !self.backgrounded else { return }
      self.deadlines[id] = nil
      self.onExpire?(id, "timeout")
    }
  }

  private func cancelTask(_ id: String) {
    tasks[id]?.cancel()
    tasks[id] = nil
  }
}

File: android/src/main/kotlin/com/rehmatsg/liquid_toasts/DeadlineScheduler.kt
package com.rehmatsg.liquid_toasts

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Owns all auto-dismiss timing for the toast stack: wall-clock deadlines (so
 * backgrounding can't corrupt them), the coroutines that fire them, and the
 * banked remaining time for toasts paused mid-interaction. A line-for-line port
 * of `DeadlineScheduler.swift`.
 *
 * Deliberately knows nothing about the stack itself — the manager guards
 * existence and reacts to [onExpire]. Flutter-free and Android-UI-free by design,
 * so it unit-tests under `runTest` virtual time with a fake [clock].
 *
 * @param scope the coroutine scope the watcher jobs run on (main-confined in
 *   production; a test scope in tests). All methods must be called on that scope's
 *   thread — this class is not internally synchronized (mirrors `@MainActor`).
 * @param clock wall-clock time in milliseconds; injectable so tests control it.
 */
internal class DeadlineScheduler(
    private val scope: CoroutineScope,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    /**
     * Fired when a deadline expires. `reason` is `"timeout"` (timer fired live)
     * or `"appBackgrounded"` (found past-due while foregrounding).
     */
    var onExpire: ((id: String, reason: String) -> Unit)? = null

    /** Absolute wall-clock deadlines (ms). Survive backgrounding. */
    private val deadlines = HashMap<String, Long>()
    private val jobs = HashMap<String, Job>()

    /** Banked remaining time (ms) for toasts paused mid-interaction (touch-down). */
    private val pausedRemaining = HashMap<String, Long>()
    private var backgrounded = false

    /**
     * Arms (or re-arms) the auto-dismiss for [id]. A null [durationMs] disarms the
     * timer. A fresh arm supersedes any banked pause.
     */
    fun arm(id: String, durationMs: Long?) {
        cancelJob(id)
        pausedRemaining.remove(id)
        if (durationMs == null) {
            deadlines.remove(id)
            return
        }
        val deadline = clock() + durationMs
        deadlines[id] = deadline
        if (!backgrounded) schedule(id, deadline)
    }

    /** Cancels everything known about [id] (teardown, async-action spinner). */
    fun disarm(id: String) {
        cancelJob(id)
        deadlines.remove(id)
        pausedRemaining.remove(id)
    }

    /**
     * Pauses [id]'s auto-dismiss while the user is touching it: banks the
     * remaining time and clears the wall-clock deadline so neither the timer nor a
     * background/foreground cycle can fire mid-interaction. No-op when there is no
     * live deadline (persistent / loading toasts, or an already-paused toast).
     */
    fun pause(id: String) {
        val deadline = deadlines[id] ?: return
        val remaining = deadline - clock()
        if (remaining <= 0) return
        pausedRemaining[id] = remaining
        cancelJob(id)
        deadlines.remove(id)
    }

    /** Resumes a paused [id] with its banked remaining time. No-op if not paused. */
    fun resume(id: String) {
        val remaining = pausedRemaining.remove(id) ?: return
        val deadline = clock() + remaining
        deadlines[id] = deadline
        if (!backgrounded) schedule(id, deadline)
    }

    /** Hot restart / flush: drop all timing state. */
    fun disarmAll() {
        for (job in jobs.values) job.cancel()
        jobs.clear()
        deadlines.clear()
        pausedRemaining.clear()
    }

    // --- App lifecycle ---

    /**
     * Cancels the watcher jobs but keeps the wall-clock deadlines — they are
     * re-evaluated against real time on foreground.
     */
    fun appDidEnterBackground() {
        backgrounded = true
        for (job in jobs.values) job.cancel()
        jobs.clear()
    }

    /**
     * Sweeps the stored deadlines: past-due ones expire immediately (reason
     * `"appBackgrounded"`), live ones get fresh watcher jobs.
     */
    fun appWillEnterForeground() {
        backgrounded = false
        val now = clock()
        // Snapshot: onExpire may mutate `deadlines` (teardown disarms).
        val snapshot = deadlines.toList()
        for ((id, deadline) in snapshot) {
            if (deadline <= now) {
                deadlines.remove(id)
                onExpire?.invoke(id, "appBackgrounded")
            } else {
                schedule(id, deadline)
            }
        }
    }

    // --- Internals ---

    private fun schedule(id: String, fireAt: Long) {
        jobs[id]?.cancel()
        // Capture the delay interval synchronously (against the current clock)
        // rather than inside the coroutine: with a lazily-started test dispatcher
        // the body would otherwise read the clock only after virtual time has
        // already advanced, collapsing the interval.
        val interval = (fireAt - clock()).coerceAtLeast(0)
        jobs[id] = scope.launch {
            delay(interval)
            // Guard against a stale fire: the job may have been cancelled (disarm /
            // background) or the app backgrounded between wake-up and this check.
            if (!isActive || backgrounded) return@launch
            deadlines.remove(id)
            onExpire?.invoke(id, "timeout")
        }
    }

    private fun cancelJob(id: String) {
        jobs.remove(id)?.cancel()
    }
}
